torch/csrc/distributed/rpc/rref_context.cpp

```diff
@@ -368,14 +367,13 @@ c10::intrusive_ptr<OwnerRRef> RRefContext::getOrCreateOwnerRRef(
 }
 
 c10::intrusive_ptr<OwnerRRef> RRefContext::createOwnerRRef(
-    const TypePtr& type,
-    std::vector<c10::Device> devices) {
+    const TypePtr& type) {
   // Don't add this OnwerRRef to the owners_ map yet, otherwise
   // it will never be removed from there. Instead, only add it to the
   // map in prepareChildFork, in case this local RRef is being passed
   // to another worker.
   return c10::make_intrusive<OwnerRRef>(
-      getWorkerId(), genGloballyUniqueId(), type, std::move(devices));
+      getWorkerId(), genGloballyUniqueId(), type, agent_->getDevices());
```

Comment:
This means RRef will always sync CUDA streams even if users just want to use RRef for CPU values, and will always need to inspect values wrapped by the RRef? Will there be perf concerns? If so, shall we try some micro-benchmark (say use RRef to wrap a cpu-only ResNet-152)?

Reply:
Right, that's a good point. I must admit, one part of me thinks we should just bite the bullet and always extract DataPtrs, even in CPU-only mode, and pay whatever perf price, since by doing this we would catch much more easily those silent CUDA sync correctness issues (e.g., all the ones I've fixed in #57432). I'm suggesting we do it not just for RRefs but for all Futures in general. In other words, correctness outweighs performance. However this only matters when the code contains bugs: in case of correct code, we'd pay this price for nothing, and that's a bummer.

I will try to do some benchmarks of this so we can get a feel of how much perf we'd be losing. I also think TorchScript here could help substantially (as inspecting an IValue is cheaper than inspecting a Python object), hence I wonder if we could have TorchScript become our "silver bullet" for performance or if it has too many feature gaps to be widely usable in RPC?

Reply:
> I wonder if we could have TorchScript become our "silver bullet" for performance

This might be hard. Based on feedback from users, some of them would like to avoid the type/API restrictions in TorchScript, even if that means they will need to take the perf hit.